idea/src/hev-segmentSelection/hev-segmentSelection.1
.TH hev-segmentSelection 1

\# This is a comment
\" This is also a comment.
\# Use .br for a line break


.SH NAME
.B hev-segmentSelection 
- Interactive specification and modification of line segments.

.SH SYNOPSIS

.B hev-segmentSelection  
[ options ]


.SH DESCRIPTION

.B hev-segmentSelection 
enables the interactive specification and modification of
line segments.

.TP
Creating a line segment:
Click left button at first point; you get a
rubberband line, then click at the second point.
Note that on the desktop, a "click" is a press/release
cycle of the virtual button.

.TP
Editing a line segment:
When the 3D cursor is close to an end point or
the segment center point, you see highlighting.
When an object is highlighted, you can press and hold the
left button, then drag the selected segment end point. 
The line will stretch as you move the endpoint.

.TP
Deleting a line segment:
When the 3D cursor is close to a line segment and
the highlighting is visible, you can click with
the right button to delete the entire segment.
A second click with the right button is needed to
confirm the delection.
Again, please note that 
on the desktop, a "click" is a press/release
cycle of the virtual button.

.PP
There is no control panel GUI for this program; it can
be controlled by sending text commands either to stdin
or through a fifo.  This is described below.

This program writes IRIS control commands to
standard output, so it is expected that this output 
will be redirected to $IRIS_CONTROL_FIFO or used in
some equivalent way.

It is important to note that coordinate systems for input
and outputs must be consistent.  See options --inCoordShm,
--ptShm, and --outNode.


.SH OPTIONS

There are appropriate defaults so that
the program can be run without any command line options
under some circumstances.

.TP
--verbose
.br
Turn on verbose mode.

.TP
--help                     
.br
Show usage message and exit.

.TP
--active
.br
Make the program active on startup.
Default is inactive.

.TP
--ptShm  outgoingPointsShmName
.br
Name of the shared memory file where segment
geometry is written.  Default is
\fBhev-segmentSelection/segmentData\fR.

.TP
--buttons  editButShmName delButShmName
.br
Shared memory files for the edit and delete buttons.
Defaults are
\fBidea/buttons/left idea/buttons/right\fR.






.TP
--fifo
.br
Receive commands via a fifo named
/dev/shm/hev-segSelCmd-$USER with line
length of 200.  If this is not specified,
then commands are received on stdin.


.TP
--selectorShm  shmName
.br
Specifies the name of a shared memory file that can be used
to activate and deactivate the operation of this program.
If this is specified, the program monitors the contents of
the indicated shared memory file.  When the string contained
in that file changes so that it matches the
\fBselectorStr\fR (see below), the program is activated.
When it changes so that they do not match, the program is deactivated.
Default is \fBidea/selector\fR.


.TP
--selectorStr  string
.br
Specifies the string that will activate this program when
present in the shared memory file specified by the
\fBselectorShm\fR
option (see above).
The default is \fBhev-segmentSelection\fR.






.TP
--inCoordShm  trackShmName
.br
Name of the the shared memory segment
from which to receive the coordinates
used for determining segment positions;
Default is \fBidea/worldOffsetWand\fR.  




.TP
--outNode   nodeName
.br
Specifies the node in the scene graph under which
all of the segment glyphs are placed.
Default is \fBworld\fR.


.TP
--maxSegments n
.br
Maximum number of line segments to allow.
Must be less than or equal to 200. Default is 200.

.TP
--lenLabels [fmt]
.br
Turns on the display of the segment
length in the virtual world; default is off
If \fBfmt\fR is present, it specifies how the length label is 
formatted; default is \fB%g\fR.

.TP
--coordScale scale
.br
Scale factor for calculated segment coordinates.
Default is 1.0.


.TP
--usleep t
.br
Specifies  how  many  microseconds  to  sleep at the end of each
loop.  If not specified, iris::GetUsleep() is used.

.br
.br


.PP
The following options specify the glyphs used at the end
points of the the segment, the middle of the segment, and
the object for the line segment itself.  If not specified,
there will be default objects that are created on the fly.
One unhighlighted form and three highlighted forms are required
for each type of glyph.

.TP
--ptMarkerFN ptFN  hl1PtFN hl2PtFN hl3PtFN
.br
The glyphs used for marking the end points
of line segments.  These should be centered at
the origin. The default glyphs are about 0.04
units across.

.TP
--lnFN  lineFN hl1LineFN  hl2LineFN  hl3LineFN
.br
The glyphs used for depicting the line segments.
These should display a line segment with one
end at the (0,0,0) and one end at (1,0,0).
The default glyphs are simple cylinders that
are about 0.004 units in diameter.

.TP
--lnHandleFN handleFN hl1HandleFN hl2HandleFN hl3HandleFN
.br
The glyphs placed at the center of each line
segment.  These should be centered at the origin.
The default glyphs are about 0.012 units across.

.TP
--glyphScale factor
.br
Scale the glyphs by the indicated factor. This scaling
is applied to the default glyphs as well as to the glyphs
provided by the options above. Default is 1.0.


.PP
Distances for highlighting and interactive selection of end points
and line segments:

.TP
--ptDist distance
.br
Distance used for grabbing and highlighting
segment end points. Default is 0.02*glyphScale.

.TP
--lnDist distance
.br
Distance from the line segment center handle that
is used for grabbing and highlighting the segment.
Default is 0.02*glyphScale.

.PP
The line segments produced by this program are written to the
shared memory file hev-segmentSelection/segmentData or as
specified by the command line option --ptShm.  Currently, the
best documentation of the format of this shared memory file
is the code itself.


.SH COMMAND INPUT

Commands can be sent to the this program either through its
standard input or through a fifo. (See the --fifo option.)
These commands can take the following form:

.TP
ACTIVATE      
.br
Enable the program to create and edit
line segments.

.TP
DEACTIVATE    
.br
Disable the program; all user interactions
are ignored.

.TP
QUIT          
.br
Terminate the program.

.TP
READ file     
.br
Read segment data from file and display.

.TP
WRITE file     
.br
Write segment data from file and display.

.TP
USERHL segID ON|OFF
.br
Turn on and off "user highlighting" for the
indicated segment.  "segID" is either the
index of an existing segment or the keyword "ALL".


.PP
Note that there are two mechanisms for activating and deactivating this program:
through the commands described in this section or through the 
shared memory method specified by the command line options
\fB--selectorShm\fR   and \fB--selectorStr\fR.
It is recommended that you do not mix these two methods.


.SH FUTURE DEVELOPMENT

Currently, all future development is anticipated to be
accomplished by adding new commands rather than adding new
options.  The biggest area of future development is the
implementation of methods that enable the user to specify
geometric constraints on the line segments.

Future Commands:

.RS

.TP
HIDE          
.br
Hide all line segments and deactivate

.TP
SHOW          
.br
Show all line segments

.TP
SCALE_DRAG factor
.br
Scale motion during drag operations.


.TP
LEN_LABEL  ON|OFF
.br
Turn on and off the display off segment lengths

.TP
RESET         
.br
Remove all existing line segments.

.TP
LIMIT  xmin ymin zmin   xmax ymax zmax
.br
Limit interactions to a box.

.TP
UNLIMIT       
.br
Remove limits

.TP
CONSTRAIN PARA x y z
.br
Constrain segments to be parallel to vector.

.TP
CONSTRAIN PERP x y z
.br
Constrain segments to be perpendicular to vector.

.TP
CONSTRAIN LENGTH l
.br
Constrain segments to be of specified length.

.TP
CONSTRAIN ONLINE  x1 y1 z1  x2 y2 z2
.br
Constrain segments to be on the specified line.

.TP
CONSTRAIN ONPLANE a b c d
.br
Constrain segments to be on the specified plane.

.TP
UNCONSTRAIN
.br
Remove constraints.

.TP
RECONSTRAIN
.br
Apply current constraints to all existing segments.

.PP
Note that CONSTRAIN commands affect future segment specification and
modification; RECONSTRAIN will alter all existing
segments to conform to the current constraints.

We will need to deal with how multiple constraints either
override or co-exist with one another.

.RE






.SH OTHER NOTES

Temporary files are written to a directory named
.br
     /var/tmp/hev-segmentSelection-$USER
.br
where the value of the environment variable USER is
substituted for "$USER".

.SH SEE ALSO
irisfly-select, irisfly-deselect




idea/src/hev-segmentSelection/hev-segmentSelection.cpp
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <dtk.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <unistd.h>
#include <errno.h>

#include <iris.h>

#include "hssOps.h"
#include "hssDBHL.h"
#include "hssCmd.h"
#include "hssFSM.h"

#include "hev-segmentSelection.h"




/*


Things to do:

	cmd line option to do calc input and output coord transformations
		either continuously or only once at the start.

	don't send DCS transforms to sge if they are unchanged

	route all error messages through a single wrapper to fprintf (stderr,

	test input & output coord node calculations

	insert verbose output

	data structures for end points and segment handles
	(Maybe make a class with access methods to keep the in-memory data
	 and the shared mem data in sync.)
	

	implement fsm

	complete usage message


*/




bool Verbose = false;
bool Debug = false;

char ProgName[1000];

char TmpDirName[1000];

static bool ActivateOnStartup = false;


static double GlyphScale = 1.0;

static void
usage ()
	{


fprintf (stderr, 
"\n"
"Usage:   hev-segmentSelection  [ options ]\n"
"\n"
"    Options:\n"
"        --verbose\n"
"        --help\n"
"        --active\n"
"        --ptShm        outPointsShmName  \n"
"                            (default: hev-segmentSelection/segmentData)\n"
"        --buttons      editButShmName delButShmName  \n"
"                            (default: idea/buttons/left idea/buttons/right)\n"
"        --fifo\n"
"        --selectorShm  shmName              (default: idea/selector)\n"
"        --selectorStr  string               (default: hev-segmentSelection)\n"
"        --inCoordShm   trackShmName         (default: idea/worldOffsetWand)\n"
"        --outNode      nodeName             (default: world)\n"
"        --maxSegments  n                    (default: 200)\n"
"        --lenLabels    [fmt]                (default: no labels)\n"
"        --coordScale   scale                (default: 1)\n"
"        --glyphScale   factor               (default: 1)\n"
"        --ptDist       distance             (default: 0.02)\n"
"        --lnDist       distance             (default: 0.02)\n"
"        --usleep       t                    (default: iris::GetUsleep()\n"
"        --ptMarkerFN   ptFN  hl1PtFN hl2PtFN hl3PtFN\n"
"        --lnFN         lineFN hl1LineFN  hl2LineFN  hl3LineFN\n"
"        --lnHandleFN   handleFN hl1HandleFN hl2HandleFN hl3HandleFN\n"
"\n"
"    For more information, see the man page.\n"
"\n"
);
	} // end of usage


static void
badArgValue (char *optionName, char *val)
	{
	fprintf (stderr, "%s: Option \"%s\" had bad value \"%s\".\n",
		ProgName, optionName, val);
	}  // end of badArgValue

static int
extractOption ( 
  int argc, 
  char **argv, 
  int iArg, 
  int *argMap, 
  int minVals, 
  int maxVals,
  int &nVals, 
  int &startVal)

	{
	if (argMap != NULL)
		{
		argMap[iArg] = 1;
		}

	startVal = iArg + 1;
	nVals = 0;


	for (int i = 0; i < maxVals; i++)
		{

		// the variable tooFew signals whether there are fewer
		// than maxVals found

		// first check if we're going beyond the end of the arg array
		bool tooFew = (startVal+i) >= argc;
		if ( argMap != NULL )
			{
			// and check if the next arg has already been taken
			// as an option
		    	tooFew |= (argMap[iArg+1+i] == 1);
		    	}

		// check if next arg starts with "--"
		tooFew |= (strncmp (argv[startVal+i], "--", 2) == 0);

		if (tooFew )
			{
			if (i < minVals)
				{
				// if there are fewer than minVals found
				fprintf (stderr, 
					"%s: Error parsing option \"%s\"; "
					"not enough values.\n", 
					ProgName, argv[iArg]);
				return -1;
				}
			break;
			}


		nVals++;
		if (argMap)
			{
			argMap[iArg+1+i] = 1;
			}
		}


	return 1;
	}  // end of extractOption





int 
getOption (
  int argc, 
  char **argv, 
  int *argMap, 
  char *name, 
  int minNameLen, 
  int minNumArgs, 
  int maxNumArgs,
  int &nArgs, 
  int &startArg)
	{
	// return 0 if the argument was not matched
	// return 1 if the argument was matched 
	// return -1 on error


	for (int i = 0; i < argc; i++)
		{

		bool checkArg ;

		if (argMap != NULL)
			{
			checkArg = ! argMap[i];
			}
		else
			{
			checkArg = 1;
			}

		if (checkArg)
			{
			// if argMap says this arg is unused

			if (strncmp (argv[i], "--", 2) == 0)
				{
				// first two chars are --


				int len = strlen (argv[i]+2);

				if ( (len >= minNameLen) && 
						(len <= strlen (name) ) )
					{
					// length of arg is within range

					if (strncasecmp (argv[i]+2, 
						              name, len) == 0)
						{
						// name matches

						if (argMap)
							{
							argMap[i] = 1;
							}

						return extractOption (
							argc, 
							argv, 
							i, 
							argMap, 
							minNumArgs, 
							maxNumArgs,
							nArgs, 
							startArg);

						}
					}
				}
			}
		}

		
	return 0;
	}  // end of getOption



static int
housekeepingArgs (int argc, char **argv, int *argMap)
	{
	int rtn;
	int nVals, startVal;
	int maxSeg;
	

	rtn = getOption ( argc, argv, argMap, 
			  const_cast<char*>("verbose"), 1, 0, 0, nVals, startVal);
	if (rtn < 0)
		{
		return -1;
		}
	else if (rtn > 0)
		{
		Verbose = true;
		// sendSgeMsg ("ECHO ON\n");
		}



	rtn = getOption ( argc, argv, argMap, 
			  const_cast<char*>("debug"), 1, 0, 0, nVals, startVal);
	if (rtn < 0)
		{
		return -1;
		}
	else if (rtn > 0)
		{
		Debug = true;
		sendSgeMsg ("ECHO ON\n");
		}





	rtn = getOption ( argc, argv, argMap, 
			  const_cast<char*>("help"), 1, 0, 0, nVals, startVal);
	if (rtn < 0)
		{
		return -1;
		}
	else if (rtn > 0)
		{
		usage ();
		EXIT (0);
		}

	rtn = getOption ( argc, argv, argMap, 
			  const_cast<char*>("active"), 1, 0, 0, nVals, startVal);
	if (rtn < 0)
		{
		return -1;
		}
	else if (rtn > 0)
		{
		ActivateOnStartup = true;
		}


	rtn = getOption ( argc, argv, argMap, 
			  const_cast<char*>("maxSegments"), 3, 1, 1, nVals, startVal);
	if (rtn < 0)
		{
		return -1;
		}
	else if (rtn > 0)
		{
		if (sscanf (argv[startVal], "%d", &maxSeg) != 1)
			{
			badArgValue (argv[startVal-1], argv[startVal]);
			return -1;
			}
		}
	else
		{
		maxSeg = 200;
		}

	setMaxSegments (maxSeg);


	rtn = getOption ( argc, argv, argMap, 
			  const_cast<char*>("usleep"), 3, 1, 1, nVals, startVal);
        unsigned long uslp;
	if (rtn < 0)
		{
		return -1;
		}
	else if (rtn > 0)
		{
                int u;
		if (sscanf (argv[startVal], "%d", &u) != 1)
			{
			badArgValue (argv[startVal-1], argv[startVal]);
			return -1;
			}
                uslp = u;
		}
	else
                {
                uslp = iris::GetUsleep();
                }

        setSleep (uslp);


	return 0;
	} // end of housekeepingArgs


static int
shmAndNodeArgs (int argc, char **argv, int *argMap)
	{
	  char *ptsShmName = const_cast<char*>("hev-segmentSelection/segmentData");
	  char *editButShmName = const_cast<char*>("idea/buttons/left");
	  char *delButShmName = const_cast<char*>("idea/buttons/right");
	  char *inCoordShmName = const_cast<char*>("idea/worldOffsetWand");
	  char *outNode = const_cast<char*>("world");

	int rtn;
	int nVal, startVal;

// fprintf(stderr, "argc = %d  argv[13] %s (%d)  argv[14] (%d) %s\n", argc,
// argv[13], argMap[13], argv[14], argMap[14] );

	rtn = getOption ( argc, argv, argMap, 
			  const_cast<char*>("ptShm"), 3, 1, 1, nVal, startVal);
	if (rtn < 0)
		{
		return -1;
		}
	else if (rtn > 0)
		{
		ptsShmName = argv[startVal];
		}



	rtn = getOption ( argc, argv, argMap, 
			  const_cast<char*>("outNode"), 4, 1, 1, nVal, startVal);
	if (rtn < 0)
		{
		return -1;
		}
	else if (rtn > 0)
		{
		outNode = argv[startVal];
		}
        else
                {
                // if arg outNode not found then try outCoordNode
	        rtn = getOption ( argc, argv, argMap, 
				  const_cast<char*>("outCoordNode"), 8, 1, 1, nVal, startVal);
	        if (rtn < 0)
		        {
		        return -1;
		        }
	        else if (rtn > 0)
		        {
		        outNode = argv[startVal];
		        }
                }

	setupPtsShm (ptsShmName, outNode);


	rtn = getOption ( argc, argv, argMap, 
			  const_cast<char*>("buttons"), 3, 2, 2, nVal, startVal);
	if (rtn < 0)
		{
		return -1;
		}
	else if (rtn > 0)
		{
		editButShmName = argv[startVal];
		delButShmName = argv[startVal+1];
		}


	if (setupButtons (editButShmName, delButShmName))
		{
		return -1;
		}


	rtn = getOption ( argc, argv, argMap, 
			  const_cast<char*>("inCoordShm"), 8, 1, 1, nVal, startVal);
	if (rtn < 0)
		{
		return -1;
		}
	else if (rtn > 0)
		{
		inCoordShmName = argv[startVal];
		}


	
	if (setupInShm (inCoordShmName))
		{
		return -1;
		}


	rtn = getOption ( argc, argv, argMap, 
			  const_cast<char*>("fifo"), 4, 0, 0, nVal, startVal);
	if (rtn < 0)
		{
		return -1;
		}
	else if (rtn > 0)
		{
		setupCmdFifo ();
		}


        char *inCtrlShm = const_cast<char*>("idea/selector");
        char *ctrlID = const_cast<char*>("hev-segmentSelection");


        rtn = getOption ( argc, argv, argMap,
			  const_cast<char*>("selectorShm"), 10, 1, 1, nVal, startVal);
        if (rtn < 0)
                {
                return -1;
                }
        else if (rtn > 0)
                {
                inCtrlShm = argv[startVal];

                }


        rtn = getOption ( argc, argv, argMap,
			  const_cast<char*>("selectorStr"), 10, 1, 1, nVal, startVal);
        if (rtn < 0)
                {
                return -1;
                }
        else if (rtn > 0)
                {
                ctrlID = argv[startVal];
                }



        if (setupCtrlShm (inCtrlShm, ctrlID))
                {
                return -1;
                }




	return 0;
	}  // end of shmAndNodeArgs



static int
objFileArgs (int argc, char **argv, int *argMap)
	{
	char objFNames[3][4][1000];
	int rtn;
	int nVal, startVal;
	double glyphScale = 1.0;
	float yOffset = 0.08;


	for (int i = 0; i < 3; i++)
		{
		for (int j = 0; j < 4; j++)
			{
			objFNames[i][j][0] = 0;
			}
		}

	rtn = getOption ( argc, argv, argMap, 
			  const_cast<char*>("glyphScale"), 3, 1, 1, nVal, startVal);
	if (rtn < 0)
		{
		return -1;
		}
	else if (rtn > 0)
		{
		if (sscanf (argv[startVal], "%lf", &glyphScale) != 1)
			{
			badArgValue (argv[startVal-1], argv[startVal]);
			return -1;
			}
        GlyphScale = glyphScale;
		}



#if 0
	rtn = getOption ( argc, argv, argMap, 
			  const_cast<char*>("offset"), 4, 1, 1, nVal, startVal);
	if (rtn < 0)
		{
		return -1;
		}
	else if (rtn > 0)
		{
		if (sscanf (argv[startVal], "%f", &yOffset) != 1)
			{
			badArgValue (argv[startVal-1], argv[startVal]);
			return -1;
			}
		}

	setYOffset (yOffset);
#endif




	rtn = getOption ( argc, argv, argMap, 
			  const_cast<char*>("ptMarkerFN"), 3, 4, 4, nVal, startVal);
	if (rtn < 0)
		{
		return -1;
		}
	else if (rtn > 0)
		{
		strcpy (objFNames[0][0], argv[startVal]);
		strcpy (objFNames[0][1], argv[startVal+1]);
		strcpy (objFNames[0][2], argv[startVal+2]);
		strcpy (objFNames[0][3], argv[startVal+3]);
		}

	rtn = getOption ( argc, argv, argMap, 
			  const_cast<char*>("lnFN"), 3, 4, 4, nVal, startVal);
	if (rtn < 0)
		{
		return -1;
		}
	else if (rtn > 0)
		{
		strcpy (objFNames[1][0], argv[startVal]);
		strcpy (objFNames[1][1], argv[startVal+1]);
		strcpy (objFNames[1][2], argv[startVal+2]);
		strcpy (objFNames[1][3], argv[startVal+3]);
		}


	rtn = getOption ( argc, argv, argMap, 
			  const_cast<char*>("lnHandleFN"), 3, 4, 4, nVal, startVal);
	if (rtn < 0)
		{
		return -1;
		}
	else if (rtn > 0)
		{
		strcpy (objFNames[2][0], argv[startVal]);
		strcpy (objFNames[2][1], argv[startVal+1]);
		strcpy (objFNames[2][2], argv[startVal+2]);
		strcpy (objFNames[2][3], argv[startVal+3]);
		}

	return setupObjects (objFNames, glyphScale);
	}  // end of objFileArgs

static int
operationalArgs (int argc, char **argv, int *argMap)
	{
	bool doLenLabel = false;
        double coordScale = 1;
	double ptDist = 0.02 * GlyphScale;
	double lnDist = 0.02 * GlyphScale;
	int rtn;
	int nVal, startVal;
        char lenFmt[100] = "%g";

	rtn = getOption ( argc, argv, argMap, 
			  const_cast<char*>("lenLabels"), 2, 0, 1, nVal, startVal);
	if (rtn < 0)
		{
		return -1;
		}
	else if (rtn > 0)
		{
		doLenLabel = true;
                if (nVal == 1)
                        {
                        strcpy (lenFmt, argv[startVal]);
                        }
		}
	displayLenLabels (doLenLabel, lenFmt);


	rtn = getOption ( argc, argv, argMap, 
			  const_cast<char*>("coordScale"), 2, 1, 1, nVal, startVal);
	if (rtn < 0)
		{
		return -1;
		}
	else if (rtn > 0)
		{
                if (sscanf (argv[startVal], "%lf", &coordScale) != 1)
                        {
			badArgValue (argv[startVal-1], argv[startVal]);
                        coordScale = 1;
                        }
		}
	setCoordScale (coordScale);


	rtn = getOption ( argc, argv, argMap, 
			  const_cast<char*>("ptDist"), 3, 1, 1, nVal, startVal);
	if (rtn < 0)
		{
		return -1;
		}
	else if (rtn > 0)
		{
		if (sscanf (argv[startVal], "%lf", &ptDist) != 1)
			{
			badArgValue (argv[startVal-1], argv[startVal]);
			return -1;
			}
		}


	rtn = getOption ( argc, argv, argMap, 
			  const_cast<char*>("lnDist"), 3, 1, 1, nVal, startVal);
	if (rtn < 0)
		{
		return -1;
		}
	else if (rtn > 0)
		{
		if (sscanf (argv[startVal], "%lf", &lnDist) != 1)
			{
			badArgValue (argv[startVal-1], argv[startVal]);
			return -1;
			}
		}

	return setEditDistances (ptDist, lnDist);
	}  // end of operationalArgs


int
processCmdLineArgs (int argc, char **argv)
	{

	int *argMap = (int *) malloc (argc*sizeof(int));
	if (argMap == NULL)
		{
		fprintf (stderr, "%s: Unable to allocate sufficient memory.\n",
			ProgName);
		EXIT (-1);
		}

	argMap[0] = 1;
	for (int i = 1; i < argc; i++)
		{
		argMap[i] = 0;
		}
 
	int rtn = 0;

	rtn += housekeepingArgs (argc, argv, argMap);

	rtn += shmAndNodeArgs (argc, argv, argMap);

	rtn += objFileArgs (argc, argv, argMap);

        rtn += operationalArgs (argc, argv, argMap);
	
	if (rtn)
		{
		return -1;
		}


	// Note that there are no non-option arguments
	int nUnused = 0;
	for (int i = 0; i < argc; i++)
		{
		if (argMap[i] == 0)
			{
			nUnused++;
			}
		}

	if (nUnused > 0)
		{
		fprintf (stderr, "%s: Unrecognized command line argument%s:\n",
			ProgName, (nUnused > 1) ? "s" : "");
		for (int i = 0; i < argc; i++)
			{
			if (argMap[i] == 0)
				{
				fprintf (stderr, "    \"%s\"\n", argv[i]);
				}
			}
		fprintf (stderr, "\n");
		return -1;
		}
		

	return 0;
	}  // end of processCmdLineArgs


static int
setup ()
	{
	strcpy (ProgName, "hev-segmentSelection");

	sprintf (TmpDirName, "/var/tmp/%s-", ProgName);
	char *userName = getenv ("USER");
	if (userName != NULL)
		{
		strcat (TmpDirName, userName);
		}

	struct stat buf;
	if (stat (TmpDirName, &buf) != 0)
		{
		if (mkdir (TmpDirName, 0777))
			{
			fprintf (stderr, 
		  	   "%s: Error creating temporary directory \"%s\":\n"
			   "      %s\n",
		  	ProgName, TmpDirName, strerror (errno));
			return -1;
			}
		}

	return 0;
	}  // end of setup

static int
removeTmpDir ()
	{

	if ( ! Debug )
		{

                executeCmd ("rm --force %s/*.savg %s/*.osg", 
					TmpDirName, TmpDirName);

		if (rmdir (TmpDirName))
			{
			fprintf (stderr, 
		  	   "%s: Error deleting temporary directory \"%s\":\n"
			   "      %s\n",
		  	ProgName, TmpDirName, strerror (errno));
			return -1;
			}
		}

	return 0;
	}  // end of removeTmpDir

#include <signal.h>
#include <sys/types.h>
#include <unistd.h>


static void
sigHandler (int sig)
	{
	fprintf (stderr, 
	 "\n----------> pid %d sigHandler received sig %d\n\n", getpid(), sig);
	if (sig == SIGSEGV)
		{
		fprintf (stderr, "SIGSEGV\n");
		sleep (1000);
		}
		
	}  // end of sigHandler

static void
setupSigHandlers ()
	{

	signal (SIGHUP, sigHandler);
	signal (SIGINT, sigHandler);
	signal (SIGQUIT, sigHandler);
	signal (SIGILL, sigHandler);
	signal (SIGTRAP, sigHandler);
	signal (SIGABRT, sigHandler);
	signal (SIGBUS, sigHandler);
	signal (SIGFPE, sigHandler);
	signal (SIGKILL, sigHandler);
	signal (SIGUSR1, sigHandler);
	signal (SIGSEGV, sigHandler);
	signal (SIGUSR2, sigHandler);
	signal (SIGPIPE, sigHandler);
	signal (SIGALRM, sigHandler);
	signal (SIGTERM, sigHandler);
	signal (SIGSTKFLT, sigHandler);
	signal (SIGCHLD, sigHandler);
	signal (SIGCONT, sigHandler);
	signal (SIGSTOP, sigHandler);
	signal (SIGTSTP, sigHandler);
	signal (SIGTTIN, sigHandler);
	signal (SIGTTOU, sigHandler);
	signal (SIGURG, sigHandler);
	signal (SIGXCPU, sigHandler);
	signal (SIGXFSZ, sigHandler);
	signal (SIGVTALRM, sigHandler);
	signal (SIGPROF, sigHandler);
	signal (SIGWINCH, sigHandler);
	signal (SIGPOLL, sigHandler);
	signal (SIGIO, sigHandler);
	signal (SIGPWR, sigHandler);
	signal (SIGSYS, sigHandler);
	signal (SIGUNUSED, sigHandler);

	}  // end of setupSigHandlers

static void 
atExitFunc ()
	{
	removeTmpDir ();
	if (Verbose)
		{
		fprintf (stderr, "hev-segmentSelection: About to exit.\n");
		}
	}  // end of atExitFunc

main (int argc, char **argv)

	{


        // This will cause all dtk messages to be written to stderr with
        // and appropriate prefix.
        dtkMsg.setFile (stderr);
        dtkMsg.setPreMessage ("hev-segmentSelection: ");
        // dtkMsg.add (DTKMSG_ERROR, 
        //             "This is an example of a error msg < %d >\n", 3);


	atexit (atExitFunc);
	// setupSigHandlers ();

	if (setup ())
		{
		EXIT (-1);
		}

	if (processCmdLineArgs (argc, argv))
		{
		EXIT (-1);
		}

	if (ActivateOnStartup)
		{
		activate ();
		}


#if 0
	while (1)
		{
		processCommands ();
		userInteractions ();
		usleep (100000);
		}
#else
	runFiniteStateMachine ();
#endif

	}  // end of main


